tests/crash.rs

```diff
@@ -0,0 +1,335 @@
+    sleep(Duration::from_millis(100)).await;
+
+    assert!(UnboundedReceiverStream::new(crash).next().await.is_some());
+    topology.stop().await;
+    sleep(Duration::from_millis(100)).await;
+
+    let output_lines = output_lines.await;
+    assert_eq!(num_lines, output_lines.len());
+    assert_eq!(input_lines, output_lines);
+}
+
+#[derive(Deserialize, Serialize, Debug)]
+struct ErrorSourceConfig;
+
+#[async_trait]
+#[typetag::serde(name = "tcp")]
+impl SourceConfig for ErrorSourceConfig {
+    async fn build(&self, _cx: SourceContext) -> Result<sources::Source, vector::Error> {
+        Ok(Box::pin(future::err(())))
+    }
+
+    fn outputs(&self, _global_log_namespace: LogNamespace) -> Vec<config::Output> {
+        vec![config::Output::default(config::DataType::Log)]
+    }
+
+    fn source_type(&self) -> &'static str {
+        "tcp"
+    }
+
+    fn can_acknowledge(&self) -> bool {
+        false
+    }
+}
+
+#[tokio::test]
+async fn test_source_error() {
+    let num_lines: usize = 10;
+
+    let in_addr = next_addr();
+    let out_addr = next_addr();
+
+    let mut config = config::Config::builder();
+    config.add_source(
+        "in",
+        sources::socket::SocketConfig::make_basic_tcp_config(in_addr),
+    );
+    config.add_source("error", ErrorSourceConfig);
+    config.add_sink(
+        "out",
+        &["in", "error"],
+        sinks::socket::SocketSinkConfig::make_basic_tcp_config(out_addr.to_string()),
+    );
+
+    let mut output_lines = CountReceiver::receive_lines(out_addr);
+
+    let (topology, crash) = start_topology(config.build().unwrap(), false).await;
+    // Wait for server to accept traffic
+    wait_for_tcp(in_addr).await;
+    sleep(Duration::from_millis(100)).await;
+
+    // Wait for output to connect
+    output_lines.connected().await;
+
+    let input_lines = random_lines(100).take(num_lines).collect::<Vec<_>>();
+    send_lines(in_addr, input_lines.clone()).await.unwrap();
+    sleep(Duration::from_millis(100)).await;
+
+    assert!(UnboundedReceiverStream::new(crash).next().await.is_some());
+    topology.stop().await;
+    sleep(Duration::from_millis(100)).await;
+
+    let output_lines = output_lines.await;
+    assert_eq!(num_lines, output_lines.len());
+    assert_eq!(input_lines, output_lines);
+}
+
+#[derive(Deserialize, Serialize, Debug)]
+struct PanicSourceConfig;
+
+#[async_trait]
+#[typetag::serde(name = "tcp")]
+impl SourceConfig for PanicSourceConfig {
+    async fn build(&self, _cx: SourceContext) -> Result<sources::Source, vector::Error> {
+        Ok(Box::pin(async { panic!() }))
+    }
+
+    fn outputs(&self, _global_log_namespace: LogNamespace) -> Vec<config::Output> {
+        vec![config::Output::default(config::DataType::Log)]
+    }
+
+    fn source_type(&self) -> &'static str {
+        "tcp"
+    }
+
+    fn can_acknowledge(&self) -> bool {
+        false
+    }
+}
+
+#[tokio::test]
+async fn test_source_panic() {
```

Comment:
I'm having a difficult time understanding what we're actually testing here.

Reply:
hah, I didn't add this. I think it's from a bad merge or something. I suspect this got deleted. I'll double check what's going on.

Reply:
Yep, it was moved to `topology` in https://github.com/vectordotdev/vector/pull/13251/files
I'll delete it...